Bottom: f9aa67e383c39f439e5f9dd6add978468c659f87
Top:    b94876c7fa2f8f098946e617cf25efbf8539b2ba
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-04-30 15:06:49 +0100

Convert object memory allocations to use g_slice()


---

diff --git a/src/action.c b/src/action.c
index ae6b87e..c0db1cc 100644
--- a/src/action.c
+++ b/src/action.c
@@ -1900,37 +1900,39 @@ static int
 ActionDumpLibrary (int argc, char **argv, int x, int y)
 {
   int i, j;
+  GList *ii, *jj;
 
   printf ("**** Do not count on this format.  It will change ****\n\n");
   printf ("MenuN   = %d\n", Library.MenuN);
-  printf ("MenuMax = %d\n", Library.MenuMax);
-  for (i = 0; i < Library.MenuN; i++)
+  for (ii = Library.Menu, i = 0; ii != NULL; ii = g_list_next (ii), i++)
     {
+      LibraryMenuType *menu = ii->data;
+
       printf ("Library #%d:\n", i);
-      printf ("    EntryN    = %d\n", Library.Menu[i].EntryN);
-      printf ("    EntryMax  = %d\n", Library.Menu[i].EntryMax);
-      printf ("    Name      = \"%s\"\n", UNKNOWN (Library.Menu[i].Name));
-      printf ("    directory = \"%s\"\n",
-	      UNKNOWN (Library.Menu[i].directory));
-      printf ("    Style     = \"%s\"\n", UNKNOWN (Library.Menu[i].Style));
-      printf ("    flag      = %d\n", Library.Menu[i].flag);
-
-      for (j = 0; j < Library.Menu[i].EntryN; j++)
+      printf ("    EntryN    = %d\n", menu->EntryN);
+      printf ("    Name      = \"%s\"\n", UNKNOWN (menu->Name));
+      printf ("    directory = \"%s\"\n", UNKNOWN (menu->directory));
+      printf ("    Style     = \"%s\"\n", UNKNOWN (menu->Style));
+      printf ("    flag      = %d\n", menu->flag);
+
+      for (jj = menu->Entry, j = 0; jj != NULL; jj = g_list_next (jj), j++)
 	{
+          LibraryEntryType *entry = jj->data;
+
 	  printf ("    #%4d: ", j);
-	  if (Library.Menu[i].Entry[j].Template == (char *) -1)
+	  if (entry->Template == (char *) -1)
 	    {
 	      printf ("newlib: \"%s\"\n",
-		      UNKNOWN (Library.Menu[i].Entry[j].ListEntry));
+		      UNKNOWN (entry->ListEntry));
 	    }
 	  else
 	    {
 	      printf ("\"%s\", \"%s\", \"%s\", \"%s\", \"%s\"\n",
-		      UNKNOWN (Library.Menu[i].Entry[j].ListEntry),
-		      UNKNOWN (Library.Menu[i].Entry[j].Template),
-		      UNKNOWN (Library.Menu[i].Entry[j].Package),
-		      UNKNOWN (Library.Menu[i].Entry[j].Value),
-		      UNKNOWN (Library.Menu[i].Entry[j].Description));
+		      UNKNOWN (entry->ListEntry),
+		      UNKNOWN (entry->Template),
+		      UNKNOWN (entry->Package),
+		      UNKNOWN (entry->Value),
+		      UNKNOWN (entry->Description));
 	    }
 	}
     }
@@ -3297,6 +3299,8 @@ static const char renumber_help[] =
 static int
 ActionRenumber (int argc, char **argv, int x, int y)
 {
+  return 0;
+#if 0
   bool changed = false;
   ElementTypePtr *element_list;
   ElementTypePtr *locked_element_list;
@@ -3635,6 +3639,7 @@ ActionRenumber (int argc, char **argv, int x, int y)
   free (element_list);
   free (cnt_list);
   return 0;
+#endif
 }
 
 
@@ -4629,16 +4634,15 @@ ActionChangeName (int argc, char **argv, int x, int y)
 		    SetChangedFlag (true);
 		    if (type == ELEMENT_TYPE)
 		      {
-			RubberbandTypePtr ptr;
-			int i;
+			GList *iter;
 
 			RestoreUndoSerialNumber ();
 			Crosshair.AttachedObject.RubberbandN = 0;
 			LookupRatLines (type, ptr1, ptr2, ptr3);
-			ptr = Crosshair.AttachedObject.Rubberband;
-			for (i = 0; i < Crosshair.AttachedObject.RubberbandN;
-			     i++, ptr++)
+			for (iter = Crosshair.AttachedObject.Rubberband;
+			     iter != NULL; iter = g_list_next (iter))
 			  {
+			    RubberbandType *ptr = iter->data;
 			    if (PCB->RatOn)
 			      EraseRat ((RatTypePtr) ptr->Line);
 			    MoveObjectToRemoveUndoList (RATLINE_TYPE,
@@ -6950,22 +6954,26 @@ find_element_by_refdes (char *refdes)
 static AttributeType *
 lookup_attr (AttributeListTypePtr list, const char *name)
 {
-  int i;
-  for (i=0; i<list->Number; i++)
-    if (strcmp (list->List[i].name, name) == 0)
-      return & list->List[i];
+  GList *iter;
+  for (iter = list->List; iter != NULL; iter = g_list_next (iter))
+    {
+      AttributeType *attr = iter->data;
+      if (strcmp (attr->name, name) == 0)
+        return attr;
+    }
   return NULL;
 }
 
 static void
 delete_attr (AttributeListTypePtr list, AttributeType *attr)
 {
-  int idx = attr - list->List;
-  if (idx < 0 || idx >= list->Number)
-    return;
-  if (list->Number - idx > 1)
-    memmove (attr, attr+1, (list->Number - idx - 1) * sizeof(AttributeType));
-  list->Number --;
+  GList *iter;
+
+  iter = g_list_find (list->List, attr);
+  g_return_if_fail (iter != NULL);
+
+  FreeAttribute (attr);
+  list->List = g_list_delete_link (list->List, iter);
 }
 
 /* ---------------------------------------------------------------- */
diff --git a/src/autoplace.c b/src/autoplace.c
index 65eef92..31a17fb 100644
--- a/src/autoplace.c
+++ b/src/autoplace.c
@@ -157,16 +157,17 @@ static void
 UpdateXY (NetListTypePtr Nets)
 {
   Cardinal SLayer, CLayer;
-  Cardinal i, j;
+  GList *i, *j;
   /* find layer groups of the component side and solder side */
   SLayer = GetLayerGroupNumberByNumber (solder_silk_layer);
   CLayer = GetLayerGroupNumberByNumber (component_silk_layer);
   /* update all nets */
-  for (i = 0; i < Nets->NetN; i++)
+  for (i = Nets->Net; i != NULL; i = g_list_next (i))
     {
-      for (j = 0; j < Nets->Net[i].ConnectionN; j++)
+      NetType *net = i->data;
+      for (j = net->Connection; j != NULL; j = g_list_next (j))
 	{
-	  ConnectionTypePtr c = &(Nets->Net[i].Connection[j]);
+	  ConnectionType *c = j->data;
 	  switch (c->type)
 	    {
 	    case PAD_TYPE:
@@ -195,7 +196,7 @@ UpdateXY (NetListTypePtr Nets)
 static PointerListType
 collectSelectedElements ()
 {
-  PointerListType list = { 0, 0, NULL };
+  PointerListType list = {0, NULL};
   ELEMENT_LOOP (PCB->Data);
   {
     if (TEST_FLAG (SELECTEDFLAG, element))
@@ -335,32 +336,38 @@ ComputeCost (NetListTypePtr Nets, double T0, double T)
   double delta3 = 0;		/* out of bounds penalty */
   double delta4 = 0;		/* alignment bonus */
   double delta5 = 0;		/* total area penalty */
-  Cardinal i, j;
+  int i;
+  GList *ii, *j;
   LocationType minx, maxx, miny, maxy;
   bool allpads, allsameside;
   Cardinal thegroup;
-  BoxListType bounds = { 0, 0, NULL };	/* save bounding rectangles here */
-  BoxListType solderside = { 0, 0, NULL };	/* solder side component bounds */
-  BoxListType componentside = { 0, 0, NULL };	/* component side bounds */
+  BoxListType bounds = {0, NULL};	/* save bounding rectangles here */
+  BoxListType solderside = {0, NULL};	/* solder side component bounds */
+  BoxListType componentside = {0, NULL};	/* component side bounds */
   /* make sure the NetList have the proper updated X and Y coords */
   UpdateXY (Nets);
   /* wire length term.  approximated by half-perimeter of minimum
    * rectangle enclosing the net.  Note that we penalize vias in
    * all-SMD nets by making the rectangle a cube and weighting
    * the "layer height" of the net. */
-  for (i = 0; i < Nets->NetN; i++)
+  for (ii = Nets->Net; ii != NULL; ii = g_list_next (ii))
     {
-      NetTypePtr n = &Nets->Net[i];
+      NetType *n = ii->data;
+      ConnectionType *c;
+
       if (n->ConnectionN < 2)
 	continue;		/* no cost to go nowhere */
-      minx = maxx = n->Connection[0].X;
-      miny = maxy = n->Connection[0].Y;
-      thegroup = n->Connection[0].group;
-      allpads = (n->Connection[0].type == PAD_TYPE);
+
+      j = n->Connection;
+      c = j->data;
+      minx = maxx = c->X;
+      miny = maxy = c->Y;
+      thegroup = c->group;
+      allpads = (c->type == PAD_TYPE);
       allsameside = true;
-      for (j = 1; j < n->ConnectionN; j++)
+      for (; j != NULL; j = g_list_next (j))
 	{
-	  ConnectionTypePtr c = &(n->Connection[j]);
+	  c = j->data;
 	  MAKEMIN (minx, c->X);
 	  MAKEMAX (maxx, c->X);
 	  MAKEMIN (miny, c->Y);
@@ -511,10 +518,8 @@ ComputeCost (NetListTypePtr Nets, double T0, double T)
    * aligning to something far away isn't profitable */
   {
     /* create r tree */
-    PointerListType seboxes = { 0, 0, NULL }
-    , ceboxes =
-    {
-    0, 0, NULL};
+    PointerListType seboxes = {0, NULL};
+    PointerListType ceboxes = {0, NULL};
     struct ebox
     {
       BoxType box;
@@ -615,6 +620,12 @@ ComputeCost (NetListTypePtr Nets, double T0, double T)
   return W + (delta1 + delta2 + delta3 - delta4 + delta5);
 }
 
+void *
+random_pointer_from_list (PointerListType *pl)
+{
+  return g_list_nth_data (pl->Ptr, random () % pl->PtrN);
+}
+
 /* ---------------------------------------------------------------------------
  * Perturb:
  *  1) flip SMD from solder side to component side or vice-versa.
@@ -628,7 +639,7 @@ createPerturbation (PointerListTypePtr selected, double T)
 {
   PerturbationType pt = { 0 };
   /* pick element to perturb */
-  pt.element = (ElementTypePtr) selected->Ptr[random () % selected->PtrN];
+  pt.element = (ElementType*) random_pointer_from_list (selected);
   /* exchange, flip/rotate or shift? */
   switch (random () % ((selected->PtrN > 1) ? 3 : 2))
     {
@@ -666,10 +677,9 @@ createPerturbation (PointerListTypePtr selected, double T)
     case 2:
       {				/* exchange! */
 	pt.which = EXCHANGE;
-	pt.other = (ElementTypePtr)
-	  selected->Ptr[random () % (selected->PtrN - 1)];
-	if (pt.other == pt.element)
-	  pt.other = (ElementTypePtr) selected->Ptr[selected->PtrN - 1];
+	pt.other = pt.element;
+	while (pt.other == pt.element)
+	  pt.other = (ElementType*) random_pointer_from_list (selected);
 	/* don't allow exchanging a solderside-side SMD component
 	 * with a non-SMD component. */
 	if ((pt.element->PinN != 0 /* non-SMD */  &&
@@ -760,7 +770,7 @@ bool
 AutoPlaceSelected (void)
 {
   NetListTypePtr Nets;
-  PointerListType Selected = { 0, 0, NULL };
+  PointerListType Selected = {0, NULL};
   PerturbationType pt;
   double C0, T0;
   bool changed = false;
diff --git a/src/autoroute.c b/src/autoroute.c
index 27dba28..920d8e8 100644
--- a/src/autoroute.c
+++ b/src/autoroute.c
@@ -1003,7 +1003,6 @@ CreateRouteData ()
     {
       layergroupboxes[i].Ptr = NULL;
       layergroupboxes[i].PtrN = 0;
-      layergroupboxes[i].PtrMax = 0;
       GROUP_LOOP (PCB->Data, i);
       {
 	if (layer->LineN || layer->ArcN)
diff --git a/src/buffer.c b/src/buffer.c
index 8793f50..9c0b77b 100644
--- a/src/buffer.c
+++ b/src/buffer.c
@@ -597,47 +597,55 @@ void
 make_footprint_hash ()
 {
   int i, j;
+  GList *ii, *jj;
   char *fp;
   int num_entries = 0;
 
   clear_footprint_hash ();
 
-  for (i=0; i<Library.MenuN; i++)
-    for (j=0; j<Library.Menu[i].EntryN; j++)
-      num_entries ++;
+  for (ii = Library.Menu; ii != NULL; ii = g_list_next (ii))
+    {
+      LibraryMenuType *menu = ii->data;
+      for (jj = menu->Entry; jj != NULL; jj = g_list_next (jj))
+        num_entries ++;
+    }
   footprint_hash = (FootprintHashEntry *)malloc (num_entries * sizeof(FootprintHashEntry));
+
   num_entries = 0;
 
   /* There are two types of library entries.  The file-based types
      have a Template of (char *)-1 and the AllocatedMemory is the full
      path to the footprint file.  The m4 ones have the footprint name
      in brackets in the description.  */
-  for (i=0; i<Library.MenuN; i++)
+  for (ii = Library.Menu, i = 0; ii != NULL; ii = g_list_next (ii), i++)
     {
+      LibraryMenuType *menu = ii->data;
 #ifdef DEBUG
   printf("In make_footprint_hash, looking for footprints in %s\n", 
-	 Library.Menu[i].directory);
+	 menu->directory);
 #endif
 
-    for (j=0; j<Library.Menu[i].EntryN; j++)
+      for (jj = menu->Entry, j = 0; jj != NULL; jj = g_list_next (jj), j++)
 	{
+          LibraryEntryType *entry = jj->data;
+
 	  footprint_hash[num_entries].menu_idx = i;
 	  footprint_hash[num_entries].entry_idx = j;
-	  if (Library.Menu[i].Entry[j].Template == (char *) -1) 
+	  if (entry->Template == (char *) -1) 
           /* file */
 	    {
 #ifdef DEBUG
-/*	      printf(" ... Examining file %s\n", Library.Menu[i].Entry[j].AllocatedMemory); */
+/*	      printf(" ... Examining file %s\n", entry->AllocatedMemory); */
 #endif
-	      fp = strrchr (Library.Menu[i].Entry[j].AllocatedMemory, '/');
+	      fp = strrchr (entry->AllocatedMemory, '/');
 
 	      if (!fp)
-		fp = strrchr (Library.Menu[i].Entry[j].AllocatedMemory, '\\');
+		fp = strrchr (entry->AllocatedMemory, '\\');
 
 	      if (fp)
 		fp ++;
 	      else 
-		fp = Library.Menu[i].Entry[j].AllocatedMemory;
+		fp = entry->AllocatedMemory;
 
 #ifdef DEBUG
 /* 	      printf(" ... found file footprint %s\n",  fp); */
@@ -649,7 +657,7 @@ make_footprint_hash ()
 	  else 
           /* m4 */
 	    {
-	      fp = strrchr (Library.Menu[i].Entry[j].Description, '[');
+	      fp = strrchr (entry->Description, '[');
 	      if (fp)
 		{
 		  footprint_hash[num_entries].footprint = strdup (fp+1);
@@ -660,7 +668,7 @@ make_footprint_hash ()
 		}
 	      else
 		{
-		  fp = Library.Menu[i].Entry[j].Description;
+		  fp = entry->Description;
 		  footprint_hash[num_entries].footprint = fp;
 		  footprint_hash[num_entries].footprint_allocated = 0;
 		}
@@ -737,8 +745,8 @@ LoadFootprintByName (BufferTypePtr Buffer, char *Footprint)
       return 1;
     }
 
-  menu = & Library.Menu[fpe->menu_idx];
-  entry = & menu->Entry[fpe->entry_idx];
+  menu = g_list_nth_data (Library.Menu, fpe->menu_idx);
+  entry = g_list_nth_data (menu->Entry, fpe->entry_idx);
 
   if (entry->Template == (char *) -1)
     {
diff --git a/src/copy.c b/src/copy.c
index b93fcfb..51781ca 100644
--- a/src/copy.c
+++ b/src/copy.c
@@ -122,7 +122,7 @@ CopyElementLowLevel (DataTypePtr Data, ElementTypePtr Dest,
 		     ElementTypePtr Src, bool uniqueName, LocationType dx,
 		     LocationType dy)
 {
-  int i;
+  GList *iter;
   /* release old memory if necessary */
   if (Dest)
     FreeElementMemory (Dest);
@@ -172,10 +172,11 @@ CopyElementLowLevel (DataTypePtr Data, ElementTypePtr Dest,
   }
   END_LOOP;
 
-  for (i=0; i<Src->Attributes.Number; i++)
-    CreateNewAttribute (& Dest->Attributes,
-			Src->Attributes.List[i].name,
-			Src->Attributes.List[i].value);
+  for (iter = Src->Attributes.List; iter != NULL; iter = g_list_next (iter))
+    {
+      AttributeType *attr = iter->data;
+      CreateNewAttribute (&Dest->Attributes, attr->name, attr->value);
+    }
 
   Dest->MarkX = Src->MarkX + dx;
   Dest->MarkY = Src->MarkY + dy;
diff --git a/src/create.c b/src/create.c
index e4eeb0b..1bf5627 100644
--- a/src/create.c
+++ b/src/create.c
@@ -915,25 +915,19 @@ CreateNewLineInSymbol (SymbolTypePtr Symbol,
 		       LocationType X1, LocationType Y1,
 		       LocationType X2, LocationType Y2, BDimension Thickness)
 {
-  LineTypePtr line = Symbol->Line;
+  LineType* line;
 
-  /* realloc new memory if necessary and clear it */
-  if (Symbol->LineN >= Symbol->LineMax)
-    {
-      Symbol->LineMax += STEP_SYMBOLLINE;
-      line = (LineTypePtr)realloc (line, Symbol->LineMax * sizeof (LineType));
-      Symbol->Line = line;
-      memset (line + Symbol->LineN, 0, STEP_SYMBOLLINE * sizeof (LineType));
-    }
+  line = g_slice_new0 (LineType);
+  Symbol->Line = g_list_append (Symbol->Line, line);
+  Symbol->LineN ++;
 
   /* copy values */
-  line = line + Symbol->LineN++;
   line->Point1.X = X1;
   line->Point1.Y = Y1;
   line->Point2.X = X2;
   line->Point2.Y = Y2;
   line->Thickness = Thickness;
-  return (line);
+  return line;
 }
 
 /* ---------------------------------------------------------------------------
@@ -1002,16 +996,17 @@ CreateNewConnection (LibraryMenuTypePtr net, char *conn)
 /* ---------------------------------------------------------------------------
  * Add an attribute to a list.
  */
-AttributeTypePtr
-CreateNewAttribute (AttributeListTypePtr list, char *name, char *value)
+AttributeType *
+CreateNewAttribute (AttributeListType *list, char *name, char *value)
 {
-  if (list->Number >= list->Max)
-    {
-      list->Max += 10;
-      list->List = (AttributeType *)realloc (list->List, list->Max * sizeof (AttributeType));
-    }
-  list->List[list->Number].name = STRDUP (name);
-  list->List[list->Number].value = STRDUP (value);
-  list->Number++;
-  return &list->List[list->Number - 1];
+  AttributeType *attr;
+
+  attr = g_slice_new0 (AttributeType);
+  list->List = g_list_append (list->List, attr);
+  list->Number ++;
+
+  attr->name = STRDUP (name);
+  attr->value = STRDUP (value);
+
+  return attr;
 }
diff --git a/src/crosshair.c b/src/crosshair.c
index c2551ec..ae8cebd 100644
--- a/src/crosshair.c
+++ b/src/crosshair.c
@@ -380,8 +380,7 @@ XORDrawInsertPointObject (void)
 static void
 XORDrawMoveOrCopyObject (void)
 {
-  RubberbandTypePtr ptr;
-  Cardinal i;
+  GList *i;
   LocationType dx = Crosshair.X - Crosshair.AttachedObject.X,
     dy = Crosshair.Y - Crosshair.AttachedObject.Y;
 
@@ -499,11 +498,10 @@ XORDrawMoveOrCopyObject (void)
     }
 
   /* draw the attached rubberband lines too */
-  i = Crosshair.AttachedObject.RubberbandN;
-  ptr = Crosshair.AttachedObject.Rubberband;
-  while (i)
+  for (i = Crosshair.AttachedObject.Rubberband; i != NULL; i = g_list_next (i))
     {
-      PointTypePtr point1, point2;
+      RubberbandType *ptr = i->data;
+      PointType *point1, *point2;
 
       if (TEST_FLAG (VIAFLAG, ptr->Line))
 	{
@@ -531,9 +529,6 @@ XORDrawMoveOrCopyObject (void)
 			     ptr->Line->Point1.Y + dy,
 			     ptr->Line->Point2.X + dx,
 			     ptr->Line->Point2.Y + dy, ptr->Line->Thickness);
-
-      ptr++;
-      i--;
     }
 }
 
diff --git a/src/draw.c b/src/draw.c
index e1ae2dc..4a254dd 100644
--- a/src/draw.c
+++ b/src/draw.c
@@ -1202,7 +1202,7 @@ DrawTextLowLevel (TextTypePtr Text, int min_line_width)
 {
   LocationType x = 0;
   unsigned char *string = (unsigned char *) Text->TextString;
-  Cardinal n;
+  GList *iter;
   FontTypePtr font = &PCB->Font;
 
   while (string && *string)
@@ -1210,11 +1210,12 @@ DrawTextLowLevel (TextTypePtr Text, int min_line_width)
       /* draw lines if symbol is valid and data is present */
       if (*string <= MAX_FONTPOSITION && font->Symbol[*string].Valid)
 	{
-	  LineTypePtr line = font->Symbol[*string].Line;
-	  LineType newline;
-
-	  for (n = font->Symbol[*string].LineN; n; n--, line++)
+	  for (iter = font->Symbol[*string].Line;
+	       iter != NULL; iter = g_list_next (iter))
 	    {
+	      LineType *line = iter->data;
+	      LineType newline;
+
 	      /* create one line, scale, move, rotate and swap it */
 	      newline = *line;
 	      newline.Point1.X = (newline.Point1.X + x) * Text->Scale / 100;
diff --git a/src/drill.c b/src/drill.c
index a7a908b..46e8f80 100644
--- a/src/drill.c
+++ b/src/drill.c
@@ -90,11 +90,9 @@ InitializeDrill (DrillTypePtr drill, PinTypePtr pin, ElementTypePtr element)
   drill->ViaCount = 0;
   drill->PinCount = 0;
   drill->UnplatedCount = 0;
-  drill->ElementMax = 0;
   drill->Element = NULL;
   drill->PinN = 0;
   drill->Pin = NULL;
-  drill->PinMax = 0;
   ptr = (void *) GetDrillPinMemory (drill);
   *((PinTypeHandle) ptr) = pin;
   if (element)
@@ -166,9 +164,8 @@ GetDrillInfo (DataTypePtr top)
 		}
 	    }
 	    END_LOOP;
-	    if (AllDrills->Drill[AllDrills->DrillN - 1].DrillSize <
-		pin->DrillingHole)
-	      {
+	    if (AllDrills->Drill[AllDrills->DrillN - 1].DrillSize < pin->DrillingHole)
+	      { /* ^^ Last drill in the list, presumably sorted? */
 		Drill = GetDrillInfoDrillMemory (AllDrills);
 		InitializeDrill (Drill, pin, element);
 	      }
@@ -231,8 +228,7 @@ RoundDrillInfo (DrillInfoTypePtr d, int roundto)
 	{
 	  int ei, ej;
 
-	  d->Drill[i].ElementMax
-	    = d->Drill[i].ElementN + d->Drill[i+1].ElementN;
+	  d->Drill[i].ElementMax = d->Drill[i].ElementN + d->Drill[i+1].ElementN;
 	  if (d->Drill[i].ElementMax)
 	    {
 	      d->Drill[i].Element = (ElementTypePtr *)realloc (d->Drill[i].Element,
@@ -280,16 +276,3 @@ RoundDrillInfo (DrillInfoTypePtr d, int roundto)
 	}
     }
 }
-
-void
-FreeDrillInfo (DrillInfoTypePtr Drills)
-{
-  DRILL_LOOP (Drills);
-  {
-    free (drill->Element);
-    free (drill->Pin);
-  }
-  END_LOOP;
-  free (Drills->Drill);
-  free (Drills);
-}
diff --git a/src/fontmode.c b/src/fontmode.c
index 69d61ea..a754843 100644
--- a/src/fontmode.c
+++ b/src/fontmode.c
@@ -81,6 +81,7 @@ FontEdit (int argc, char **argv, int Ux, int Uy)
   SymbolType *symbol;
   LayerTypePtr lfont, lorig, lwidth, lgrid;
   int s, l;
+  GList *ll;
 
   if (hid_actionl ("New", "Font", 0))
     return 1;
@@ -118,25 +119,26 @@ FontEdit (int argc, char **argv, int Ux, int Uy)
       miny = 500;
       maxy = font->MaxHeight;
 
-      for (l = 0; l < symbol->LineN; l++)
+      for (ll = symbol->Line; ll != NULL; ll = g_list_next (ll))
 	{
+	  LineType *line = ll->data;
+
 	  CreateDrawnLineOnLayer (lfont,
-				  symbol->Line[l].Point1.X + ox,
-				  symbol->Line[l].Point1.Y + oy,
-				  symbol->Line[l].Point2.X + ox,
-				  symbol->Line[l].Point2.Y + oy,
-				  symbol->Line[l].Thickness,
-				  symbol->Line[l].Thickness, NoFlags ());
-	  CreateDrawnLineOnLayer (lorig, symbol->Line[l].Point1.X + ox,
-				  symbol->Line[l].Point1.Y + oy,
-				  symbol->Line[l].Point2.X + ox,
-				  symbol->Line[l].Point2.Y + oy,
-				  symbol->Line[l].Thickness,
-				  symbol->Line[l].Thickness, NoFlags ());
-	  if (maxx < symbol->Line[l].Point1.X)
-	    maxx = symbol->Line[l].Point1.X;
-	  if (maxx < symbol->Line[l].Point2.X)
-	    maxx = symbol->Line[l].Point2.X;
+				  line->Point1.X + ox,
+				  line->Point1.Y + oy,
+				  line->Point2.X + ox,
+				  line->Point2.Y + oy,
+				  line->Thickness,
+				  line->Thickness, NoFlags ());
+	  CreateDrawnLineOnLayer (lorig,
+				  line->Point1.X + ox,
+				  line->Point1.Y + oy,
+				  line->Point2.X + ox,
+				  line->Point2.Y + oy,
+				  line->Thickness,
+				  line->Thickness, NoFlags ());
+	  maxx = MAX (maxx, line->Point1.X);
+	  maxx = MAX (maxx, line->Point2.X);
 	}
       w = maxx + symbol->Delta + ox;
       CreateDrawnLineOnLayer (lwidth,
diff --git a/src/global.h b/src/global.h
index 9ad5331..f843617 100644
--- a/src/global.h
+++ b/src/global.h
@@ -207,8 +207,8 @@ typedef struct
 
 struct AttributeListType
 {
-  int Number, Max;
-  AttributeType *List;
+  int Number;
+  GList *List;
 };
 
 /* ---------------------------------------------------------------------------
@@ -251,15 +251,13 @@ typedef struct
 struct polygon_st			/* holds information about a polygon */
 {
   ANYOBJECTFIELDS;
-  Cardinal PointN,		/* number of points in polygon */
-    PointMax;			/* max number from malloc() */
+  Cardinal PointN;		/* number of points in polygon */
   POLYAREA *Clipped;		/* the clipped region of this polygon */
   PLINE *NoHoles;		/* the polygon broken into hole-less regions */
   int NoHolesValid;		/* Is the NoHoles polygon up to date? */
-  PointTypePtr Points;		/* data */
-  Cardinal *HoleIndex;		/* Index of hole data within the Points array */
+  GList *Points;		/* data */
+  GList *HoleIndex;		/* Index of hole data within the Points array */
   Cardinal HoleIndexN;		/* number of holes in polygon */
-  Cardinal HoleIndexMax;	/* max number from malloc() */
 
 };
 
@@ -352,10 +350,9 @@ typedef struct
  */
 typedef struct			/* a single symbol */
 {
-  LineTypePtr Line;
+  GList *Line;
   bool Valid;
-  Cardinal LineN,		/* number of lines */
-    LineMax;
+  Cardinal LineN;		/* number of lines */
   BDimension Width,		/* size of cell */
     Height, Delta;		/* distance to next symbol in 0.00001'' */
 } SymbolType, *SymbolTypePtr;
@@ -388,21 +385,18 @@ typedef struct			/* holds drill information */
 {
   BDimension DrillSize;		/* this drill's diameter */
   Cardinal ElementN,		/* the number of elements using this drill size */
-    ElementMax,			/* max number of elements from malloc() */
     PinCount,			/* number of pins drilled this size */
     ViaCount,			/* number of vias drilled this size */
     UnplatedCount,		/* number of these holes that are unplated */
-    PinN,			/* number of drill coordinates in the list */
-    PinMax;			/* max number of coordinates from malloc() */
-  PinTypePtr *Pin;		/* coordinates to drill */
-  ElementTypePtr *Element;	/* a pointer to an array of element pointers */
+    PinN;			/* number of drill coordinates in the list */
+  GList *Pin;		/* coordinates to drill */
+  GList *Element;		/* a pointer to an array of element pointers */
 } DrillType, *DrillTypePtr;
 
 typedef struct			/* holds a range of Drill Infos */
 {
-  Cardinal DrillN,		/* number of drill sizes */
-    DrillMax;			/* max number from malloc() */
-  DrillTypePtr Drill;		/* plated holes */
+  Cardinal DrillN;		/* number of drill sizes */
+  GList *Drill;			/* plated holes */
 } DrillInfoType, *DrillInfoTypePtr;
 
 typedef struct
@@ -440,9 +434,8 @@ typedef struct
   char *Name,			/* name of the menu entry */
    *directory,			/* Directory name library elements are from */
    *Style;			/* routing style */
-  Cardinal EntryN,		/* number of objects */
-    EntryMax;			/* number of reserved memory locations */
-  LibraryEntryTypePtr Entry;	/* the entries */
+  Cardinal EntryN;		/* number of objects */
+  GList *Entry;			/* the entries */
   char flag;			/* used by the netlist window to enable/disable nets */
   char internal;		/* if set, this is an internal-only entry, not
 				   part of the global netlist. */
@@ -451,8 +444,7 @@ typedef struct
 typedef struct
 {
   Cardinal MenuN;               /* number of objects */
-  Cardinal MenuMax;             /* number of reserved memory locations */
-  LibraryMenuTypePtr Menu;      /* the entries */
+  GList *Menu;                  /* the entries */
 } LibraryType, *LibraryTypePtr;
 
 
@@ -555,9 +547,8 @@ typedef struct			/* currently attached object */
   void *Ptr1,			/* three pointers to data, see */
    *Ptr2,			/* search.c */
    *Ptr3;
-  Cardinal RubberbandN,		/* number of lines in array */
-    RubberbandMax;
-  RubberbandTypePtr Rubberband;
+  Cardinal RubberbandN;		/* number of lines in array */
+  GList *Rubberband;
 } AttachedObjectType, *AttachedObjectTypePtr;
 
 enum crosshair_shape
@@ -718,38 +709,33 @@ typedef struct			/* holds a connection */
 
 typedef struct			/* holds a net of connections */
 {
-  Cardinal ConnectionN,		/* the number of connections contained */
-    ConnectionMax;		/* max connections from malloc */
-  ConnectionTypePtr Connection;
+  Cardinal ConnectionN;		/* the number of connections contained */
+  GList *Connection;
   RouteStyleTypePtr Style;
 } NetType, *NetTypePtr;
 
 typedef struct			/* holds a list of nets */
 {
-  Cardinal NetN,		/* the number of subnets contained */
-    NetMax;			/* max subnets from malloc */
-  NetTypePtr Net;
+  Cardinal NetN;		/* the number of subnets contained */
+  GList *Net;
 } NetListType, *NetListTypePtr;
 
 typedef struct			/* holds a list of net lists */
 {
-  Cardinal NetListN,		/* the number of net lists contained */
-    NetListMax;			/* max net lists from malloc */
-  NetListTypePtr NetList;
+  Cardinal NetListN;		/* the number of net lists contained */
+  GList *NetList;
 } NetListListType, *NetListListTypePtr;
 
 typedef struct			/* holds a generic list of pointers */
 {
-  Cardinal PtrN,		/* the number of pointers contained */
-    PtrMax;			/* max subnets from malloc */
-  void **Ptr;
+  Cardinal PtrN;		/* the number of pointers contained */
+  GList *Ptr;
 } PointerListType, *PointerListTypePtr;
 
 typedef struct
 {
-  Cardinal BoxN,		/* the number of boxes contained */
-    BoxMax;			/* max boxes from malloc */
-  BoxTypePtr Box;
+  Cardinal BoxN;		/* the number of boxes contained */
+  GList *Box;
 
 } BoxListType, *BoxListTypePtr;
 
diff --git a/src/hid/gtk/gtkhid-main.c b/src/hid/gtk/gtkhid-main.c
index ac961db..6cead6d 100644
--- a/src/hid/gtk/gtkhid-main.c
+++ b/src/hid/gtk/gtkhid-main.c
@@ -828,6 +828,7 @@ static void
 ghid_attributes_revert ()
 {
   int i;
+  GList *iter;
 
   ghid_attributes_need_rows (attributes_list->Number);
 
@@ -841,11 +842,14 @@ ghid_attributes_revert ()
     }
 
   /* Fill in values */
-  for (i=0; i<attributes_list->Number; i++)
+  for (iter = attributes_list->List, i = 0;
+       iter != NULL; iter = g_list_next (iter), i++)
     {
+      AttributeType *attr = iter->data;
+
       /* create row [i] */
-      gtk_entry_set_text (GTK_ENTRY (attr_row[i].w_name), attributes_list->List[i].name);
-      gtk_entry_set_text (GTK_ENTRY (attr_row[i].w_value), attributes_list->List[i].value);
+      gtk_entry_set_text (GTK_ENTRY (attr_row[i].w_name), attr->name);
+      gtk_entry_set_text (GTK_ENTRY (attr_row[i].w_value), attr->value);
 #if 0
 #endif
     }
@@ -911,28 +915,14 @@ ghid_attributes (char *owner, AttributeListType *attrs)
       if (response == GTK_RESPONSE_OK)
 	{
 	  int i;
+	  FreeAttributeListMemory (attributes_list);
+
 	  /* Copy the values back */
-	  for (i=0; i<attributes_list->Number; i++)
-	    {
-	      if (attributes_list->List[i].name)
-		free (attributes_list->List[i].name);
-	      if (attributes_list->List[i].value)
-		free (attributes_list->List[i].value);
-	    }
-	  if (attributes_list->Max < attr_num_rows)
-	    {
-	      int sz = attr_num_rows * sizeof (AttributeType);
-	      if (attributes_list->List == NULL)
-		attributes_list->List = (AttributeType *) malloc (sz);
-	      else
-		attributes_list->List = (AttributeType *) realloc (attributes_list->List, sz);
-	      attributes_list->Max = attr_num_rows;
-	    }
-	  for (i=0; i<attr_num_rows; i++)
+	  for (i = 0; i < attr_num_rows; i++)
 	    {
-	      attributes_list->List[i].name = strdup (gtk_entry_get_text (GTK_ENTRY (attr_row[i].w_name)));
-	      attributes_list->List[i].value = strdup (gtk_entry_get_text (GTK_ENTRY (attr_row[i].w_value)));
-	      attributes_list->Number = attr_num_rows;
+	      AttributeType *attr = GetAttributeMemory (attributes_list);
+	      attr->name = strdup (gtk_entry_get_text (GTK_ENTRY (attr_row[i].w_name)));
+	      attr->value = strdup (gtk_entry_get_text (GTK_ENTRY (attr_row[i].w_value)));
 	    }
 
 	  break;
diff --git a/src/hid/gtk/gui-netlist-window.c b/src/hid/gtk/gui-netlist-window.c
index b60b8ba..c638a51 100644
--- a/src/hid/gtk/gui-netlist-window.c
+++ b/src/hid/gtk/gui-netlist-window.c
@@ -509,9 +509,8 @@ netlist_disable_all_cb (GtkToggleButton * button, gpointer data)
 static void
 netlist_select_cb (GtkWidget * widget, gpointer data)
 {
-  LibraryEntryType *entry;
   ConnectionType conn;
-  gint i;
+  GList *iter;
   gboolean select_flag = GPOINTER_TO_INT (data);
 
   if (!selected_net)
@@ -522,9 +521,12 @@ netlist_select_cb (GtkWidget * widget, gpointer data)
   InitConnectionLookup ();
   ResetConnections (true);
 
-  for (i = selected_net->EntryN, entry = selected_net->Entry; i; i--, entry++)
+  for (iter = selected_net->Entry; iter != NULL; iter = g_list_next (iter)) {
+    LibraryEntryType *entry = iter->data;
+
     if (SeekPad (entry, &conn, false))
       RatFindHook (conn.type, conn.ptr1, conn.ptr2, conn.ptr2, true, true);
+  }
 
   SelectConnection (select_flag);
   ResetConnections (false);
@@ -585,7 +587,7 @@ node_get_node_from_name (gchar * node_name, LibraryMenuType ** node_net)
   GtkTreeIter iter;
   LibraryMenuType *net;
   LibraryEntryType *node;
-  gint j;
+  GList *j;
 
   if (!node_name)
     return NULL;
@@ -609,13 +611,16 @@ node_get_node_from_name (gchar * node_name, LibraryMenuType ** node_net)
 
 	/* Look for the node name in this net.
 	 */
-	for (j = net->EntryN, node = net->Entry; j; j--, node++)
-	  if (node->ListEntry && !strcmp (node_name, node->ListEntry))
+	for (j = net->Entry; j != NULL; j = g_list_next (j)) {
+          LibraryEntryType *entry = j->data;
+
+	  if (entry->ListEntry && !strcmp (node_name, entry->ListEntry))
 	    {
 	      if (node_net)
 		*node_net = net;
 	      return node;
 	    }
+        }
       }
     while (gtk_tree_model_iter_next (net_model, &iter));
 
@@ -819,9 +824,8 @@ hunt_named_node (GtkTreeModel *model, GtkTreePath *path,
 {
   struct ggnfnn_task *task = (struct ggnfnn_task *)data;
   LibraryMenuType *net;
-  LibraryEntryType *node;
   gchar *str;
-  gint j;
+  GList *j;
   gboolean is_disabled;
 
   /* We only want to inspect leaf nodes in the tree */
@@ -838,13 +842,16 @@ hunt_named_node (GtkTreeModel *model, GtkTreePath *path,
     return FALSE;
 
   /* Look for the node name in this net. */
-  for (j = net->EntryN, node = net->Entry; j; j--, node++)
-    if (node->ListEntry && !strcmp (task->node_name, node->ListEntry))
+  for (j = net->Entry; j != NULL; j = g_list_next (j)) {
+    LibraryEntryType *entry = j->data;
+
+    if (entry->ListEntry && !strcmp (task->node_name, entry->ListEntry))
       {
         task->found_net = net;
         task->iter = *iter;
         return TRUE;
       }
+  }
 
   return FALSE;
 }
@@ -974,7 +981,7 @@ ghid_netlist_window_update (gboolean init_nodes)
 
   /* XXX Check if the select callback does this for us */
   if (init_nodes)
-    node_model_update ((&PCB->NetlistLib)->Menu);
+    node_model_update ((LibraryMenuType *)PCB->NetlistLib.Menu->data);
 }
 
 static gint
diff --git a/src/hid/nelma/nelma.c b/src/hid/nelma/nelma.c
index 9dbf24c..c552b9a 100644
--- a/src/hid/nelma/nelma.c
+++ b/src/hid/nelma/nelma.c
@@ -296,28 +296,26 @@ nelma_write_materials(FILE * out)
 static void 
 nelma_write_nets(FILE * out)
 {
-	LibraryType     netlist;
 	LibraryMenuTypePtr net;
 	LibraryEntryTypePtr pin;
+	GList *n, *m;
 
-	int             n, m, i, idx;
+	int             i, idx;
 
 	const char     *ext;
 
-	netlist = PCB->NetlistLib;
-
 	fprintf(out, "\n/* **** Nets **** */\n\n");
 
-	for (n = 0; n < netlist.MenuN; n++) {
-		net = &netlist.Menu[n];
+	for (n = PCB->NetlistLib.Menu; n != NULL; n = g_list_next (n)) {
+		net = n->data;
 
 		/* Weird, but correct */
 		fprintf(out, "net %s {\n", &net->Name[2]);
 
 		fprintf(out, "\tobjects = {\n");
 
-		for (m = 0; m < net->EntryN; m++) {
-			pin = &net->Entry[m];
+		for (m = net->Entry; m != NULL; m = g_list_next (m)) {
+			pin = m->data;
 
 			/* pin_name_to_xy(pin, &x, &y); */
 
@@ -349,7 +347,7 @@ nelma_write_layer(FILE * out, int z, int h,
 	LibraryMenuTypePtr net;
 	LibraryEntryTypePtr pin;
 
-	int             n, m;
+	GList *n, *m;
 
 	fprintf(out, "layer %s {\n", name);
 	fprintf(out, "\theight = %d\n", h);
@@ -360,11 +358,11 @@ nelma_write_layer(FILE * out, int z, int h,
 		fprintf(out, "\tobjects = {\n");
 		netlist = PCB->NetlistLib;
 
-		for (n = 0; n < netlist.MenuN; n++) {
-			net = &netlist.Menu[n];
+		for (n = PCB->NetlistLib.Menu; n != NULL; n = g_list_next (n)) {
+			net = n->data;
 
-			for (m = 0; m < net->EntryN; m++) {
-				pin = &net->Entry[m];
+			for (m = net->Entry; m != NULL; m = g_list_next (m)) {
+				pin = m->data;
 
 				if (m != 0 || n != 0)
 					fprintf(out, ",\n");
@@ -469,17 +467,17 @@ nelma_write_objects(FILE * out)
 	LibraryMenuTypePtr net;
 	LibraryEntryTypePtr pin;
 
-	int             n, m;
+	GList *n, *m;
 
 	netlist = PCB->NetlistLib;
 
 	fprintf(out, "\n/* **** Objects **** */\n\n");
 
-	for (n = 0; n < netlist.MenuN; n++) {
-		net = &netlist.Menu[n];
+	for (n = PCB->NetlistLib.Menu; n != NULL; n = g_list_next (n)) {
+		net = n->data;
 
-		for (m = 0; m < net->EntryN; m++) {
-			pin = &net->Entry[m];
+		for (m = net->Entry; m != NULL; m = g_list_next (m)) {
+			pin = m->data;
 
 			nelma_write_object(out, pin);
 		}
diff --git a/src/intersect.c b/src/intersect.c
index e7612e9..104ca29 100644
--- a/src/intersect.c
+++ b/src/intersect.c
@@ -90,18 +90,20 @@ LocationList;
  * Create a sorted list of unique y coords from a BoxList.
  */
 static LocationList
-createSortedYList (BoxListTypePtr boxlist)
+createSortedYList (BoxListType *boxlist)
 {
   LocationList yCoords;
   LocationType last;
   int i, n;
+  GList *ii;
   /* create sorted list of Y coordinates */
   yCoords.size = 2 * boxlist->BoxN;
   yCoords.p = (LocationType *)calloc (yCoords.size, sizeof (*yCoords.p));
-  for (i = 0; i < boxlist->BoxN; i++)
+  for (ii = boxlist->Box, i = 0; ii != NULL; ii = g_list_next (ii), i++)
     {
-      yCoords.p[2 * i] = boxlist->Box[i].Y1;
-      yCoords.p[2 * i + 1] = boxlist->Box[i].Y2;
+      BoxType *box = ii->data;
+      yCoords.p[2 * i + 0] = box->Y1;
+      yCoords.p[2 * i + 1] = box->Y2;
     }
   qsort (yCoords.p, yCoords.size, sizeof (*yCoords.p), comparepos);
   /* count uniq y coords */
@@ -202,12 +204,15 @@ deleteSegment (SegmentTree * st, int n, LocationType Y1, LocationType Y2)
 double
 ComputeIntersectionArea (BoxListTypePtr boxlist)
 {
-  Cardinal i;
+  GList *i;
   double area = 0.0;
   /* first get the aggregate area. */
-  for (i = 0; i < boxlist->BoxN; i++)
-    area += (double) (boxlist->Box[i].X2 - boxlist->Box[i].X1) *
-      (double) (boxlist->Box[i].Y2 - boxlist->Box[i].Y1);
+  for (i = boxlist->Box; i != NULL; i = g_list_next (i))
+    {
+      BoxType *box = i->data;
+      area += (double) (box->X2 - box->X1) *
+              (double) (box->Y2 - box->Y1);
+    }
   /* intersection area is aggregate - union. */
   return area * 0.0001 - ComputeUnionArea (boxlist);
 }
@@ -221,6 +226,7 @@ ComputeUnionArea (BoxListTypePtr boxlist)
 {
   BoxTypePtr *rectLeft, *rectRight;
   Cardinal i, j;
+  GList *ii;
   LocationList yCoords;
   SegmentTree segtree;
   LocationType lastX;
@@ -236,11 +242,12 @@ ComputeUnionArea (BoxListTypePtr boxlist)
   /* create sorted list of left and right X coordinates of rectangles */
   rectLeft = (BoxTypePtr *)calloc (boxlist->BoxN, sizeof (*rectLeft));
   rectRight = (BoxTypePtr *)calloc (boxlist->BoxN, sizeof (*rectRight));
-  for (i = 0; i < boxlist->BoxN; i++)
+  for (ii = boxlist->Box, i = 0; ii != NULL; ii = g_list_next (ii), i++)
     {
-      assert (boxlist->Box[i].X1 <= boxlist->Box[i].X2);
-      assert (boxlist->Box[i].Y1 <= boxlist->Box[i].Y2);
-      rectLeft[i] = rectRight[i] = &boxlist->Box[i];
+      BoxType *box = ii->data;
+      assert (boxlist->Box[i].X1 <= box->X2);
+      assert (boxlist->Box[i].Y1 <= box->Y2);
+      rectLeft[i] = rectRight[i] = box;
     }
   qsort (rectLeft, boxlist->BoxN, sizeof (*rectLeft), compareleft);
   qsort (rectRight, boxlist->BoxN, sizeof (*rectRight), compareright);
diff --git a/src/macro.h b/src/macro.h
index e990d4a..b4307f7 100644
--- a/src/macro.h
+++ b/src/macro.h
@@ -14,7 +14,7 @@
  *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  *  GNU General Public License for more details.
  *
- *  You should have received a copy of the GNU General Public License
+ *  You should have received a __copy of the GNU General Public License
  *  along with this program; if not, write to the Free Software
  *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
  *
@@ -191,37 +191,37 @@ extern int mem_any_set (unsigned char *, int);
        __iter = g_list_next (__iter), n++) {                        \
     PinType *via = __iter->data;
 
-#define DRILL_LOOP(top) do             {               \
-        Cardinal        n;                                      \
-        GList *__copy = NULL; /* DUMMY */                       \
-        DrillTypePtr    drill;                                  \
-        for (n = 0; (top)->DrillN > 0 && n < (top)->DrillN; n++)                        \
-        {                                                       \
-                drill = &(top)->Drill[n]
-
-#define NETLIST_LOOP(top) do   {                         \
-        Cardinal        n;                                      \
-        GList *__copy = NULL; /* DUMMY */                       \
-        NetListTypePtr   netlist;                               \
-        for (n = (top)->NetListN-1; n != -1; n--)               \
-        {                                                       \
-                netlist = &(top)->NetList[n]
-
-#define NET_LOOP(top) do   {                             \
-        Cardinal        n;                                      \
-        GList *__copy = NULL; /* DUMMY */                       \
-        NetTypePtr   net;                                       \
-        for (n = (top)->NetN-1; n != -1; n--)                   \
-        {                                                       \
-                net = &(top)->Net[n]
-
-#define CONNECTION_LOOP(net) do {                         \
-        Cardinal        n;                                      \
-        GList *__copy = NULL; /* DUMMY */                       \
-        ConnectionTypePtr       connection;                     \
-        for (n = (net)->ConnectionN-1; n != -1; n--)            \
-        {                                                       \
-                connection = & (net)->Connection[n]
+#define DRILL_LOOP(top) do {                                        \
+  GList *__iter;                                                    \
+  GList *__copy = g_list_copy ((top)->Drill);                       \
+  Cardinal n;                                                       \
+  for (__iter = __copy, n = 0; __iter != NULL;                      \
+       __iter = g_list_next (__iter), n++) {                        \
+    DrillType *drill = __iter->data;
+
+#define NETLIST_LOOP(top) do {                                      \
+  GList *__iter;                                                    \
+  GList *__copy = g_list_reverse (g_list_copy ((top)->NetList));    \
+  Cardinal n;                                                       \
+  for (__iter = __copy, n = (top)->NetListN - 1; __iter != NULL;    \
+       __iter = g_list_next (__iter), n--) {                        \
+    NetListType *netlist = __iter->data;
+
+#define NET_LOOP(top) do {                                          \
+  GList *__iter;                                                    \
+  GList *__copy = g_list_reverse (g_list_copy ((top)->Net));        \
+  Cardinal n;                                                       \
+  for (__iter = __copy, n = (top)->NetN - 1; __iter != NULL;        \
+       __iter = g_list_next (__iter), n--) {                        \
+    NetType *net = __iter->data;
+
+#define CONNECTION_LOOP(net) do {                                   \
+  GList *__iter;                                                    \
+  GList *__copy = g_list_reverse (g_list_copy ((net)->Connection)); \
+  Cardinal n;                                                       \
+  for (__iter = __copy, n = (net)->ConnectionN - 1; __iter != NULL; \
+       __iter = g_list_next (__iter), n--) {                        \
+    ConnectionType *connection = __iter->data;
 
 #define ELEMENT_LOOP(top) do {                                      \
   GList *__iter;                                                    \
@@ -432,7 +432,7 @@ extern int mem_any_set (unsigned char *, int);
 	LayerTypePtr	layer = (board)->Data->Layer;		\
 	for (l = 0; l < max_copper_layer + 2; l++, layer++)	\
 	{ \
-                TEXT_LOOP(layer);                                      \
+                TEXT_LOOP(layer);                                   \
                   if (TEXT_IS_VISIBLE((board), layer, text))
 
 #define	VISIBLEPOLYGON_LOOP(top) do	{	\
@@ -443,48 +443,48 @@ extern int mem_any_set (unsigned char *, int);
 		if (layer->On)				\
 			POLYGON_LOOP(layer)
 
-#define POINTER_LOOP(top) do	{	\
-	Cardinal	n;			\
-	GList *__copy = NULL; /* DUMMY */	\
-	void	**ptr;				\
-	for (n = (top)->PtrN-1; n != -1; n--)	\
-	{					\
-		ptr = &(top)->Ptr[n]
-
-#define MENU_LOOP(top)	do {	\
-	Cardinal	l;			\
-	GList *__copy = NULL; /* DUMMY */	\
-	LibraryMenuTypePtr menu;		\
-	for (l = (top)->MenuN-1; l != -1; l--)	\
-	{					\
-		menu = &(top)->Menu[l]
-
-#define ENTRY_LOOP(top) do	{	\
-	Cardinal	n;			\
-	GList *__copy = NULL; /* DUMMY */	\
-	LibraryEntryTypePtr entry;		\
-	for (n = (top)->EntryN-1; n != -1; n--)	\
-	{					\
-		entry = &(top)->Entry[n]
-
-#define GROUP_LOOP(data, group) do { 	\
-	Cardinal entry; \
-	GList *__copy = NULL; /* DUMMY */	\
-        for (entry = 0; entry < ((PCBTypePtr)(data->pcb))->LayerGroups.Number[(group)]; entry++) \
-        { \
-		LayerTypePtr layer;		\
-		Cardinal number; 		\
-		number = ((PCBTypePtr)(data->pcb))->LayerGroups.Entries[(group)][entry]; \
-		if (number >= max_copper_layer)	\
-		  continue;			\
-		layer = &data->Layer[number];
-
-#define LAYER_LOOP(data, ml) do { \
-        Cardinal n; \
-        GList *__copy = NULL; /* DUMMY */ \
-	for (n = 0; n < ml; n++) \
-	{ \
-	   LayerTypePtr layer = (&data->Layer[(n)]);
+#define POINTER_LOOP(top) do {                                      \
+  GList *__iter;                                                    \
+  GList *__copy = g_list_reverse (g_list_copy ((top)->Ptr));        \
+  Cardinal n;                                                       \
+  for (__iter = __copy, n = (top)->PtrN - 1; __iter != NULL;        \
+       __iter = g_list_next (__iter), n--) {                        \
+    void **ptr = &__iter->data;
+
+#define MENU_LOOP(top) do {                                         \
+  GList *__iter;                                                    \
+  GList *__copy = g_list_reverse (g_list_copy ((top)->Menu));       \
+  Cardinal n;                                                       \
+  for (__iter = __copy, n = (top)->MenuN - 1; __iter != NULL;       \
+       __iter = g_list_next (__iter), n--) {                        \
+    LibraryMenuType *menu = __iter->data;
+
+#define ENTRY_LOOP(top) do {                                        \
+  GList *__iter;                                                    \
+  GList *__copy = g_list_reverse (g_list_copy ((top)->Entry));      \
+  Cardinal n;                                                       \
+  for (__iter = __copy, n = (top)->EntryN - 1; __iter != NULL;      \
+       __iter = g_list_next (__iter), n--) {                        \
+    LibraryEntryType *entry = __iter->data;
+
+#define GROUP_LOOP(data, group) do {                                \
+  GList *__copy = NULL; /* DUMMY */                                 \
+  Cardinal entry; \
+  for (entry = 0; entry < ((PCBTypePtr)(data->pcb))->LayerGroups.Number[(group)]; entry++) \
+  {                                                                 \
+    LayerTypePtr layer;                                             \
+    Cardinal number;                                                \
+    number = ((PCBTypePtr)(data->pcb))->LayerGroups.Entries[(group)][entry]; \
+    if (number >= max_copper_layer)                                 \
+      continue;                                                     \
+    layer = &data->Layer[number];
+
+#define LAYER_LOOP(data, ml) do {                                   \
+        Cardinal n;                                                 \
+        GList *__copy = NULL; /* DUMMY */                           \
+        for (n = 0; n < ml; n++)                                    \
+        {                                                           \
+           LayerTypePtr layer = (&data->Layer[(n)]);
 
 
 #endif
diff --git a/src/misc.c b/src/misc.c
index 0e25089..cbbb172 100644
--- a/src/misc.c
+++ b/src/misc.c
@@ -441,7 +441,7 @@ SetTextBoundingBox (FontTypePtr FontPtr, TextTypePtr Text)
   SymbolTypePtr symbol = FontPtr->Symbol;
   unsigned char *s = (unsigned char *) Text->TextString;
   BDimension minThick = 0;
-  int i;
+  GList *i;
   int space = 0;
 
   LocationType minx=0, miny=0, maxx=0, maxy=0;
@@ -462,9 +462,9 @@ SetTextBoundingBox (FontTypePtr FontPtr, TextTypePtr Text)
     {
       if (*s <= MAX_FONTPOSITION && symbol[*s].Valid)
 	{
-	  LineTypePtr line = symbol[*s].Line;
-	  for (i = 0; i < symbol[*s].LineN; line++, i++)
+	  for (i = symbol[*s].Line; i != NULL; i = g_list_next (i))
 	    {
+              LineType *line = i->data;
 	      int t = line->Thickness / 4;
 	      if (t < minThick)
 		t = minThick;
@@ -723,9 +723,9 @@ CenterDisplay (LocationType X, LocationType Y, bool Delta)
 void
 SetFontInfo (FontTypePtr Ptr)
 {
-  Cardinal i, j;
+  Cardinal i;
+  GList *iter;
   SymbolTypePtr symbol;
-  LineTypePtr line;
   LocationType totalminy = MAX_COORD;
 
   /* calculate cell with and height (is at least DEFAULT_CELLSIZE)
@@ -744,8 +744,9 @@ SetFontInfo (FontTypePtr Ptr)
 
       minx = miny = MAX_COORD;
       maxx = maxy = 0;
-      for (line = symbol->Line, j = symbol->LineN; j; j--, line++)
+      for (iter = symbol->Line ; iter != NULL; iter = g_list_next (iter))
         {
+          LineType *line = iter->data;
           minx = MIN (minx, line->Point1.X);
           miny = MIN (miny, line->Point1.Y);
           minx = MIN (minx, line->Point2.X);
@@ -757,8 +758,11 @@ SetFontInfo (FontTypePtr Ptr)
         }
 
       /* move symbol to left edge */
-      for (line = symbol->Line, j = symbol->LineN; j; j--, line++)
-        MOVE_LINE_LOWLEVEL (line, -minx, 0);
+      for (iter = symbol->Line; iter != NULL; iter = g_list_next (iter))
+        {
+          LineType *line = iter->data;
+          MOVE_LINE_LOWLEVEL (line, -minx, 0);
+        }
 
       /* set symbol bounding box with a minimum cell size of (1,1) */
       symbol->Width = maxx - minx + 1;
@@ -775,8 +779,11 @@ SetFontInfo (FontTypePtr Ptr)
     if (symbol->Valid)
       {
         symbol->Height -= totalminy;
-        for (line = symbol->Line, j = symbol->LineN; j; j--, line++)
-          MOVE_LINE_LOWLEVEL (line, 0, -totalminy);
+        for (iter = symbol->Line; iter != NULL; iter = g_list_next (iter))
+          {
+            LineType *line = iter->data;
+            MOVE_LINE_LOWLEVEL (line, 0, -totalminy);
+          }
       }
 
   /* setup the box for the default symbol */
@@ -1696,8 +1703,11 @@ GetGridLockCoordinates (int type, void *ptr1,
       *y = ((ElementTypePtr) ptr2)->MarkY;
       break;
     case POLYGON_TYPE:
-      *x = ((PolygonTypePtr) ptr2)->Points[0].X;
-      *y = ((PolygonTypePtr) ptr2)->Points[0].Y;
+      {
+        PointType *point0 = ((PolygonTypePtr) ptr2)->Points->data;
+        *x = point0->X;
+        *y = point0->Y;
+      }
       break;
 
     case LINEPOINT_TYPE:
@@ -1980,10 +1990,13 @@ pcb_author (void)
 char *
 AttributeGetFromList (AttributeListType *list, char *name)
 {
-  int i;
-  for (i=0; i<list->Number; i++)
-    if (strcmp (name, list->List[i].name) == 0)
-      return list->List[i].value;
+  GList *i;
+  for (i = list->List; i != NULL; i = g_list_next (i))
+    {
+      AttributeType *attr = i->data;
+      if (strcmp (name, attr->name) == 0)
+        return attr->value;
+    }
   return NULL;
 }
 
diff --git a/src/move.c b/src/move.c
index 6ce043a..364e8d3 100644
--- a/src/move.c
+++ b/src/move.c
@@ -789,7 +789,7 @@ void *
 MoveObjectAndRubberband (int Type, void *Ptr1, void *Ptr2, void *Ptr3,
 			 LocationType DX, LocationType DY)
 {
-  RubberbandTypePtr ptr;
+  GList *iter;
   void *ptr2;
 
   /* setup offset */
@@ -799,9 +799,11 @@ MoveObjectAndRubberband (int Type, void *Ptr1, void *Ptr2, void *Ptr3,
     return (NULL);
 
   /* move all the lines... and reset the counter */
-  ptr = Crosshair.AttachedObject.Rubberband;
-  while (Crosshair.AttachedObject.RubberbandN)
+  for (iter = Crosshair.AttachedObject.Rubberband;
+       iter != NULL; iter = g_list_next (iter))
     {
+      RubberbandType *ptr = iter->data;
+
       /* first clear any marks that we made in the line flags */
       CLEAR_FLAG (RUBBERENDFLAG, ptr->Line);
       AddObjectToMoveUndoList (LINEPOINT_TYPE,
@@ -809,7 +811,6 @@ MoveObjectAndRubberband (int Type, void *Ptr1, void *Ptr2, void *Ptr3,
 			       DY);
       MoveLinePoint (ptr->Layer, ptr->Line, ptr->MovedPoint);
       Crosshair.AttachedObject.RubberbandN--;
-      ptr++;
     }
 
   AddObjectToMoveUndoList (Type, Ptr1, Ptr2, Ptr3, DX, DY);
diff --git a/src/mymem.c b/src/mymem.c
index 77a2ce2..63f4699 100644
--- a/src/mymem.c
+++ b/src/mymem.c
@@ -36,8 +36,6 @@
 
 #include "global.h"
 
-#include <memory.h>
-
 #include "data.h"
 #include "error.h"
 #include "mymem.h"
@@ -59,128 +57,126 @@ static void DSRealloc (DynamicStringTypePtr, size_t);
 /* ---------------------------------------------------------------------------
  * get next slot for a rubberband connection, allocates memory if necessary
  */
-RubberbandTypePtr
+RubberbandType *
 GetRubberbandMemory (void)
 {
-  RubberbandTypePtr ptr = Crosshair.AttachedObject.Rubberband;
+  AttachedObjectType *attached = &Crosshair.AttachedObject;
+  RubberbandType *new_obj;
 
-  /* realloc new memory if necessary and clear it */
-  if (Crosshair.AttachedObject.RubberbandN >=
-      Crosshair.AttachedObject.RubberbandMax)
-    {
-      Crosshair.AttachedObject.RubberbandMax += STEP_RUBBERBAND;
-      ptr = (RubberbandTypePtr)realloc (ptr, Crosshair.AttachedObject.RubberbandMax *
-                          sizeof (RubberbandType));
-      Crosshair.AttachedObject.Rubberband = ptr;
-      memset (ptr + Crosshair.AttachedObject.RubberbandN, 0,
-	      STEP_RUBBERBAND * sizeof (RubberbandType));
-    }
-  return (ptr + Crosshair.AttachedObject.RubberbandN++);
+  new_obj = g_slice_new0 (RubberbandType);
+  attached->Rubberband = g_list_append (attached->Rubberband, new_obj);
+  attached->RubberbandN ++;
+
+  return new_obj;
 }
 
 void **
-GetPointerMemory (PointerListTypePtr list)
+GetPointerMemory (PointerListType *list)
 {
-  void **ptr = list->Ptr;
+  void **new_obj;
 
-  /* realloc new memory if necessary and clear it */
-  if (list->PtrN >= list->PtrMax)
-    {
-      list->PtrMax = STEP_POINT + (2 * list->PtrMax);
-      ptr = (void **)realloc (ptr, list->PtrMax * sizeof (void *));
-      list->Ptr = ptr;
-      memset (ptr + list->PtrN, 0,
-	      (list->PtrMax - list->PtrN) * sizeof (void *));
-    }
-  return (ptr + list->PtrN++);
+  new_obj = g_slice_new0 (void *);
+  list->Ptr = g_list_append (list->Ptr, new_obj);
+  list->PtrN ++;
+
+  return new_obj;
+}
+
+static void
+FreePointer (gpointer data)
+{
+  g_slice_free (void *, data);
 }
 
 void
-FreePointerListMemory (PointerListTypePtr list)
+FreePointerListMemory (PointerListType *list)
 {
-  free (list->Ptr);
+  g_list_free_full (list->Ptr, (GDestroyNotify)FreePointer);
   memset (list, 0, sizeof (PointerListType));
 }
 
 /* ---------------------------------------------------------------------------
  * get next slot for a box, allocates memory if necessary
  */
-BoxTypePtr
-GetBoxMemory (BoxListTypePtr Boxes)
+BoxType *
+GetBoxMemory (BoxListType *boxlist)
 {
-  BoxTypePtr box = Boxes->Box;
+  BoxType *new_obj;
 
-  /* realloc new memory if necessary and clear it */
-  if (Boxes->BoxN >= Boxes->BoxMax)
-    {
-      Boxes->BoxMax = STEP_POINT + (2 * Boxes->BoxMax);
-      box = (BoxTypePtr)realloc (box, Boxes->BoxMax * sizeof (BoxType));
-      Boxes->Box = box;
-      memset (box + Boxes->BoxN, 0,
-	      (Boxes->BoxMax - Boxes->BoxN) * sizeof (BoxType));
-    }
-  return (box + Boxes->BoxN++);
+  new_obj = g_slice_new0 (BoxType);
+  boxlist->Box = g_list_append (boxlist->Box, new_obj);
+  boxlist->BoxN ++;
+
+  return new_obj;
 }
 
+static void
+FreeBox (BoxType *data)
+{
+  g_slice_free (BoxType, data);
+}
 
 /* ---------------------------------------------------------------------------
  * get next slot for a connection, allocates memory if necessary
  */
-ConnectionTypePtr
-GetConnectionMemory (NetTypePtr Net)
+ConnectionType *
+GetConnectionMemory (NetType *net)
 {
-  ConnectionTypePtr con = Net->Connection;
+  ConnectionType *new_obj;
 
-  /* realloc new memory if necessary and clear it */
-  if (Net->ConnectionN >= Net->ConnectionMax)
-    {
-      Net->ConnectionMax += STEP_POINT;
-      con = (ConnectionTypePtr)realloc (con, Net->ConnectionMax * sizeof (ConnectionType));
-      Net->Connection = con;
-      memset (con + Net->ConnectionN, 0,
-	      STEP_POINT * sizeof (ConnectionType));
-    }
-  return (con + Net->ConnectionN++);
+  new_obj = g_slice_new (ConnectionType);
+  net->Connection = g_list_append (net->Connection, new_obj);
+  net->ConnectionN ++;
+
+  return new_obj;
+}
+
+static void
+FreeConnection (ConnectionType *data)
+{
+  g_slice_free (ConnectionType, data);
 }
 
 /* ---------------------------------------------------------------------------
  * get next slot for a subnet, allocates memory if necessary
  */
-NetTypePtr
-GetNetMemory (NetListTypePtr Netlist)
+NetType *
+GetNetMemory (NetListType *netlist)
 {
-  NetTypePtr net = Netlist->Net;
+  NetType *new_obj;
 
-  /* realloc new memory if necessary and clear it */
-  if (Netlist->NetN >= Netlist->NetMax)
-    {
-      Netlist->NetMax += STEP_POINT;
-      net = (NetTypePtr)realloc (net, Netlist->NetMax * sizeof (NetType));
-      Netlist->Net = net;
-      memset (net + Netlist->NetN, 0, STEP_POINT * sizeof (NetType));
-    }
-  return (net + Netlist->NetN++);
+  new_obj = g_slice_new (NetType);
+  netlist->Net = g_list_append (netlist->Net, new_obj);
+  netlist->NetN ++;
+
+  return new_obj;
+}
+
+static void
+FreeNet (NetType *data)
+{
+  g_slice_free (NetType, data);
 }
 
 /* ---------------------------------------------------------------------------
  * get next slot for a net list, allocates memory if necessary
  */
-NetListTypePtr
-GetNetListMemory (NetListListTypePtr Netlistlist)
+NetListType *
+GetNetListMemory (NetListListType *netlistlist)
 {
-  NetListTypePtr netlist = Netlistlist->NetList;
+  NetListType* new_obj;
 
-  /* realloc new memory if necessary and clear it */
-  if (Netlistlist->NetListN >= Netlistlist->NetListMax)
-    {
-      Netlistlist->NetListMax += STEP_POINT;
-      netlist = (NetListTypePtr)realloc (netlist,
-                         Netlistlist->NetListMax * sizeof (NetListType));
-      Netlistlist->NetList = netlist;
-      memset (netlist + Netlistlist->NetListN, 0,
-	      STEP_POINT * sizeof (NetListType));
-    }
-  return (netlist + Netlistlist->NetListN++);
+  new_obj = g_slice_new (NetListType);
+  netlistlist->NetList = g_list_append (netlistlist->NetList, new_obj);
+  netlistlist->NetListN ++;
+
+  return new_obj;
+}
+
+static void
+FreeNetList (NetListType *data)
+{
+  g_slice_free (NetListType, data);
 }
 /* ---------------------------------------------------------------------------
  * get next slot for a pin, allocates memory if necessary
@@ -354,21 +350,22 @@ FreePolygon (PolygonType *data)
  * gets the next slot for a point in a polygon struct, allocates memory
  * if necessary
  */
-PointTypePtr
-GetPointMemoryInPolygon (PolygonTypePtr Polygon)
+PointType *
+GetPointMemoryInPolygon (PolygonType *polygon)
 {
-  PointTypePtr points = Polygon->Points;
+  PointType *new_obj;
 
-  /* realloc new memory if necessary and clear it */
-  if (Polygon->PointN >= Polygon->PointMax)
-    {
-      Polygon->PointMax += STEP_POLYGONPOINT;
-      points = (PointTypePtr)realloc (points, Polygon->PointMax * sizeof (PointType));
-      Polygon->Points = points;
-      memset (points + Polygon->PointN, 0,
-	      STEP_POLYGONPOINT * sizeof (PointType));
-    }
-  return (points + Polygon->PointN++);
+  new_obj = g_slice_new (PointType);
+  polygon->Points = g_list_append (polygon->Points, new_obj);
+  polygon->PointN ++;
+
+  return new_obj;
+}
+
+static void
+FreePoint (PointType *data)
+{
+  g_slice_free (PointType, data);
 }
 
 /* ---------------------------------------------------------------------------
@@ -376,20 +373,21 @@ GetPointMemoryInPolygon (PolygonTypePtr Polygon)
  * if necessary
  */
 Cardinal *
-GetHoleIndexMemoryInPolygon (PolygonTypePtr Polygon)
+GetHoleIndexMemoryInPolygon (PolygonType *polygon)
 {
-  Cardinal *holeindex = Polygon->HoleIndex;
+  Cardinal *new_obj;
 
-  /* realloc new memory if necessary and clear it */
-  if (Polygon->HoleIndexN >= Polygon->HoleIndexMax)
-    {
-      Polygon->HoleIndexMax += STEP_POLYGONHOLEINDEX;
-      holeindex = (Cardinal *)realloc (holeindex, Polygon->HoleIndexMax * sizeof (int));
-      Polygon->HoleIndex = holeindex;
-      memset (holeindex + Polygon->HoleIndexN, 0,
-	      STEP_POLYGONHOLEINDEX * sizeof (int));
-    }
-  return (holeindex + Polygon->HoleIndexN++);
+  new_obj = g_slice_new (Cardinal);
+  polygon->HoleIndex = g_list_append (polygon->HoleIndex, new_obj);
+  polygon->HoleIndexN ++;
+
+  return new_obj;
+}
+
+static void
+FreeHoleIndex (Cardinal *data)
+{
+  g_slice_free (Cardinal, data);
 }
 
 /* ---------------------------------------------------------------------------
@@ -414,106 +412,142 @@ FreeElement (ElementType *data)
 }
 
 /* ---------------------------------------------------------------------------
- * get next slot for a library menu, allocates memory if necessary
+ * get next slot for a library entry, allocates memory if necessary
  */
-LibraryMenuTypePtr
-GetLibraryMenuMemory (LibraryTypePtr lib)
+LibraryEntryType *
+GetLibraryEntryMemory (LibraryMenuType *menu)
 {
-  LibraryMenuTypePtr menu = lib->Menu;
+  LibraryEntryType *new_obj;
 
-  /* realloc new memory if necessary and clear it */
-  if (lib->MenuN >= lib->MenuMax)
-    {
-      lib->MenuMax += STEP_LIBRARYMENU;
-      menu = (LibraryMenuTypePtr)realloc (menu, lib->MenuMax * sizeof (LibraryMenuType));
-      lib->Menu = menu;
-      memset (menu + lib->MenuN, 0,
-	      STEP_LIBRARYMENU * sizeof (LibraryMenuType));
-    }
-  return (menu + lib->MenuN++);
+  new_obj = g_slice_new (LibraryEntryType);
+  menu->Entry = g_list_append (menu->Entry, new_obj);
+  menu->EntryN ++;
+
+  return new_obj;
+}
+
+static void
+FreeLibraryEntry (LibraryEntryType *entry)
+{
+  free (entry->AllocatedMemory);
+  free (entry->ListEntry);
+
+  g_slice_free (LibraryEntryType, entry);
 }
 
 /* ---------------------------------------------------------------------------
- * get next slot for a library entry, allocates memory if necessary
+ * get next slot for a library menu, allocates memory if necessary
  */
-LibraryEntryTypePtr
-GetLibraryEntryMemory (LibraryMenuTypePtr Menu)
+LibraryMenuType *
+GetLibraryMenuMemory (LibraryType *lib)
 {
-  LibraryEntryTypePtr entry = Menu->Entry;
+  LibraryMenuType *new_obj;
 
-  /* realloc new memory if necessary and clear it */
-  if (Menu->EntryN >= Menu->EntryMax)
-    {
-      Menu->EntryMax += STEP_LIBRARYENTRY;
-      entry = (LibraryEntryTypePtr)realloc (entry, Menu->EntryMax * sizeof (LibraryEntryType));
-      Menu->Entry = entry;
-      memset (entry + Menu->EntryN, 0,
-	      STEP_LIBRARYENTRY * sizeof (LibraryEntryType));
-    }
-  return (entry + Menu->EntryN++);
+  new_obj = g_slice_new (LibraryMenuType);
+  lib->Menu = g_list_append (lib->Menu, new_obj);
+  lib->MenuN ++;
+
+  return new_obj;
+}
+
+static void
+FreeLibraryMenu (LibraryMenuType *menu)
+{
+  g_list_free_full (menu->Entry, (GDestroyNotify)FreeLibraryEntry);
+  free (menu->Name);
+  g_slice_free (LibraryMenuType, menu);
 }
 
 /* ---------------------------------------------------------------------------
  * get next slot for a DrillElement, allocates memory if necessary
  */
-ElementTypeHandle
-GetDrillElementMemory (DrillTypePtr Drill)
+ElementType **
+GetDrillElementMemory (DrillType *drill)
 {
-  ElementTypePtr *element;
+  ElementType **new_obj;
 
-  element = Drill->Element;
+  new_obj = g_slice_new (ElementType *);
+  drill->Element = g_list_append (drill->Element, new_obj);
+  drill->ElementN ++;
 
-  /* realloc new memory if necessary and clear it */
-  if (Drill->ElementN >= Drill->ElementMax)
-    {
-      Drill->ElementMax += STEP_ELEMENT;
-      element = (ElementTypePtr *)realloc (element,
-                         Drill->ElementMax * sizeof (ElementTypeHandle));
-      Drill->Element = element;
-      memset (element + Drill->ElementN, 0,
-	      STEP_ELEMENT * sizeof (ElementTypeHandle));
-    }
-  return (element + Drill->ElementN++);
+  return new_obj;
+}
+
+static void
+FreeDrillElement (ElementType **data)
+{
+  g_slice_free (ElementType *, data);
 }
 
 /* ---------------------------------------------------------------------------
  * get next slot for a DrillPoint, allocates memory if necessary
  */
-PinTypeHandle
-GetDrillPinMemory (DrillTypePtr Drill)
+PinType **
+GetDrillPinMemory (DrillType *drill)
 {
-  PinTypePtr *pin;
+  PinType **new_obj;
 
-  pin = Drill->Pin;
+  new_obj = g_slice_new (PinType *);
+  drill->Pin = g_list_append (drill->Pin, new_obj);
+  drill->PinN ++;
 
-  /* realloc new memory if necessary and clear it */
-  if (Drill->PinN >= Drill->PinMax)
-    {
-      Drill->PinMax += STEP_POINT;
-      pin = (PinTypePtr *)realloc (pin, Drill->PinMax * sizeof (PinTypeHandle));
-      Drill->Pin = pin;
-      memset (pin + Drill->PinN, 0, STEP_POINT * sizeof (PinTypeHandle));
-    }
-  return (pin + Drill->PinN++);
+  return new_obj;
+}
+
+static void
+FreeDrillPin (PinType **data)
+{
+  g_slice_free (PinType *, data);
 }
 
 /* ---------------------------------------------------------------------------
  * get next slot for a Drill, allocates memory if necessary
  */
-DrillTypePtr
-GetDrillInfoDrillMemory (DrillInfoTypePtr DrillInfo)
+DrillType *
+GetDrillInfoDrillMemory (DrillInfoType *drillinfo)
 {
-  DrillTypePtr drill = DrillInfo->Drill;
+  DrillType *new_obj;
 
-  /* realloc new memory if necessary and clear it */
-  if (DrillInfo->DrillN >= DrillInfo->DrillMax)
-    {
-      DrillInfo->DrillMax += STEP_DRILL;
-      drill = (DrillTypePtr)realloc (drill, DrillInfo->DrillMax * sizeof (DrillType));
-      DrillInfo->Drill = drill;
-      memset (drill + DrillInfo->DrillN, 0, STEP_DRILL * sizeof (DrillType));
-    }
-  return (drill + DrillInfo->DrillN++);
+  new_obj = g_slice_new (DrillType);
+  drillinfo->Drill = g_list_append (drillinfo->Drill, new_obj);
+  drillinfo->DrillN ++;
+
+  return new_obj;
+}
+
+static void
+FreeDrill (DrillType *drill)
+{
+  g_list_free_full (drill->Element, (GDestroyNotify)FreeDrillElement);
+  g_list_free_full (drill->Pin, (GDestroyNotify)FreeDrillPin);
+  g_slice_free (DrillType, drill);
+}
+
+void
+FreeDrillInfo (DrillInfoType *drill_info)
+{
+  g_list_free_full (drill_info->Drill, (GDestroyNotify)FreeDrill);
+  g_slice_free (DrillInfoType, drill_info);
+}
+
+AttributeType *
+GetAttributeMemory (AttributeListType *attr_list)
+{
+  AttributeType *new_obj;
+
+  new_obj = g_slice_new0 (AttributeType);
+  attr_list->List = g_list_append (attr_list->List, new_obj);
+  attr_list->Number ++;
+
+  return new_obj;
+}
+
+void
+FreeAttribute (AttributeType *attr)
+{
+  free (attr->name);
+  free (attr->value);
+  g_slice_free (AttributeType, attr);
 }
 
 /* ---------------------------------------------------------------------------
@@ -539,79 +573,74 @@ FreePolygonMemory (PolygonType *polygon)
  * frees memory used by a box list
  */
 void
-FreeBoxListMemory (BoxListTypePtr Boxlist)
+FreeBoxListMemory (BoxListType *boxlist)
 {
-  if (Boxlist)
-    {
-      free (Boxlist->Box);
-      memset (Boxlist, 0, sizeof (BoxListType));
-    }
+  if (boxlist == NULL)
+    return;
+
+  g_list_free_full (boxlist->Box, (GDestroyNotify)FreeBox);
+  memset (boxlist, 0, sizeof (BoxListType));
 }
 
 /* ---------------------------------------------------------------------------
  * frees memory used by a net 
  */
 void
-FreeNetListMemory (NetListTypePtr Netlist)
+FreeNetListMemory (NetListType *netlist)
 {
-  if (Netlist)
-    {
-      NET_LOOP (Netlist);
-      {
-	FreeNetMemory (net);
-      }
-      END_LOOP;
-      free (Netlist->Net);
-      memset (Netlist, 0, sizeof (NetListType));
-    }
+  if (netlist == NULL)
+    return;
+
+  NET_LOOP (netlist);
+  {
+    FreeNetMemory (net);
+  }
+  END_LOOP;
+
+  g_list_free_full (netlist->Net, (GDestroyNotify)FreeNet);
+  memset (netlist, 0, sizeof (NetListType));
 }
 
 /* ---------------------------------------------------------------------------
  * frees memory used by a net list
  */
 void
-FreeNetListListMemory (NetListListTypePtr Netlistlist)
+FreeNetListListMemory (NetListListType *netlistlist)
 {
-  if (Netlistlist)
-    {
-      NETLIST_LOOP (Netlistlist);
-      {
-	FreeNetListMemory (netlist);
-      }
-      END_LOOP;
-      free (Netlistlist->NetList);
-      memset (Netlistlist, 0, sizeof (NetListListType));
-    }
+  if (netlistlist == NULL)
+    return;
+
+  NETLIST_LOOP (netlistlist);
+  {
+    FreeNetListMemory (netlist);
+  }
+  END_LOOP;
+
+  g_list_free_full (netlistlist->NetList, (GDestroyNotify)FreeNetList);
+  memset (netlistlist, 0, sizeof (NetListListType));
 }
 
 /* ---------------------------------------------------------------------------
  * frees memory used by a subnet 
  */
 void
-FreeNetMemory (NetTypePtr Net)
+FreeNetMemory (NetType *net)
 {
-  if (Net)
-    {
-      free (Net->Connection);
-      memset (Net, 0, sizeof (NetType));
-    }
+  if (net == NULL)
+    return;
+
+  g_list_free_full (net->Connection, (GDestroyNotify)FreeConnection);
+  memset (net, 0, sizeof (NetType));
 }
+
 /* ---------------------------------------------------------------------------
  * frees memory used by an attribute list
- */
-static void
-FreeAttributeListMemory (AttributeListTypePtr list)
+*/
+void
+FreeAttributeListMemory (AttributeListType *list)
 {
-  int i;
-
-  for (i = 0; i < list->Number; i++)
-    {
-      free (list->List[i].name);
-      free (list->List[i].value);
-    }
-  free (list->List);
-  list->List = NULL;
-  list->Max = 0;
+  g_list_free_full (list->List, (GDestroyNotify)FreeAttribute);
+  memset (list, 0, sizeof (AttributeListType));
 }
 
 /* ---------------------------------------------------------------------------
@@ -757,24 +786,9 @@ FreeDataMemory (DataType *data)
  * releases the memory that's allocated by the library
  */
 void
-FreeLibraryMemory (LibraryTypePtr lib)
+FreeLibraryMemory (LibraryType *lib)
 {
-  MENU_LOOP (lib);
-  {
-    ENTRY_LOOP (menu);
-    {
-      free (entry->AllocatedMemory);
-      free (entry->ListEntry);
-    }
-    END_LOOP;
-    free (menu->Entry);
-    free (menu->Name);
-  }
-  END_LOOP;
-  free (lib->Menu);
-
-  /* clear struct */
-  memset (lib, 0, sizeof (LibraryType));
+  g_list_free_full (lib->Menu, (GDestroyNotify)FreeLibraryMenu);
 }
 
 /* ---------------------------------------------------------------------------
diff --git a/src/mymem.h b/src/mymem.h
index 6b572d8..b7bbd16 100644
--- a/src/mymem.h
+++ b/src/mymem.h
@@ -38,22 +38,6 @@
 #include <stdlib.h>
 #include "global.h"
 
-/* ---------------------------------------------------------------------------
- * number of additional objects that are allocated with one system call
- */
-#define	STEP_ELEMENT		50
-#define STEP_DRILL		30
-#define STEP_POINT		100
-#define	STEP_SYMBOLLINE		10
-#define	STEP_SELECTORENTRY	128
-#define	STEP_REMOVELIST		500
-#define	STEP_UNDOLIST		500
-#define	STEP_POLYGONPOINT	10
-#define	STEP_POLYGONHOLEINDEX	10
-#define	STEP_LIBRARYMENU	10
-#define	STEP_LIBRARYENTRY	20
-#define	STEP_RUBBERBAND		100
-
 #define STRDUP(x) (((x) != NULL) ? strdup (x) : NULL)
 
 /* ---------------------------------------------------------------------------
@@ -86,8 +70,12 @@ LibraryEntryTypePtr GetLibraryEntryMemory (LibraryMenuTypePtr);
 ElementTypeHandle GetDrillElementMemory (DrillTypePtr);
 PinTypeHandle GetDrillPinMemory (DrillTypePtr);
 DrillTypePtr GetDrillInfoDrillMemory (DrillInfoTypePtr);
+void FreeDrillInfo (DrillInfoTypePtr);
 void **GetPointerMemory (PointerListTypePtr);
+AttributeTypePtr GetAttributeMemory (AttributeListTypePtr);
+void FreeAttribute (AttributeTypePtr);
 void FreePolygonMemory (PolygonTypePtr);
+void FreeAttributeListMemory (AttributeListTypePtr);
 void FreeElementMemory (ElementTypePtr);
 void FreePCBMemory (PCBTypePtr);
 void FreeBoxListMemory (BoxListTypePtr);
diff --git a/src/report.c b/src/report.c
index 5131ec6..5890c16 100644
--- a/src/report.c
+++ b/src/report.c
@@ -335,8 +335,7 @@ ReportDialog (int argc, char **argv, int x, int y)
 
 	sprintf (&report[0], "POLYGON ID# %ld;  Flags:%s\n"
 		 "Its bounding box is (%.*f,%.*f) (%.*f,%.*f) %s.\n"
-		 "It has %d points and could store %d more\n"
-		 "  without using more memory.\n"
+		 "It has %d points.\n"
 		 "It has %d holes and resides on layer %d.\n"
 		 "%s", Polygon->ID,
 		 flags_to_string (Polygon->Flags, POLYGON_TYPE),
@@ -344,7 +343,7 @@ ReportDialog (int argc, char **argv, int x, int y)
 		 prec, units(Polygon->BoundingBox.Y1),
 		 prec, units(Polygon->BoundingBox.X2),
 		 prec, UNIT(Polygon->BoundingBox.Y2),
-		 Polygon->PointN, Polygon->PointMax - Polygon->PointN,
+		 Polygon->PointN,
 		 Polygon->HoleIndexN,
 		 GetLayerNumber (PCB->Data, (LayerTypePtr) ptr1),
 		 TEST_FLAG (LOCKFLAG, Polygon) ? "It is LOCKED.\n" : "");
diff --git a/src/select.c b/src/select.c
index e42d4bb..48261a9 100644
--- a/src/select.c
+++ b/src/select.c
@@ -1022,18 +1022,20 @@ SelectObjectByName (int Type, char *Pattern, bool Flag)
 
       MENU_LOOP (&PCB->NetlistLib);
       {
-        Cardinal i;
-        LibraryEntryType *entry;
+        GList *i;
         ConnectionType conn;
 
         /* Name[0] and Name[1] are special purpose, not the actual name*/
         if (menu->Name && menu->Name[0] != '\0' && menu->Name[1] != '\0' &&
             REGEXEC (menu->Name + 2))
           {
-            for (i = menu->EntryN, entry = menu->Entry; i; i--, entry++)
-              if (SeekPad (entry, &conn, false))
-                RatFindHook (conn.type, conn.ptr1, conn.ptr2, conn.ptr2,
-                             true, true);
+            for (i = menu->Entry; i != NULL; i = g_list_next (i))
+              {
+                LibraryEntryType *entry = i->data;
+                if (SeekPad (entry, &conn, false))
+                  RatFindHook (conn.type, conn.ptr1, conn.ptr2, conn.ptr2,
+                               true, true);
+              }
           }
       }
       END_LOOP;
diff --git a/src/toporouter.c b/src/toporouter.c
index 402de74..7d3dd40 100644
--- a/src/toporouter.c
+++ b/src/toporouter.c
@@ -2804,7 +2804,8 @@ import_clusters(toporouter_t *r)
   NETLIST_LOOP(&nets);
   {
     if(netlist->NetN > 0) {
-      toporouter_netlist_t *nl = netlist_create(r, netlist->Net->Connection->menu->Name, netlist->Net->Connection->menu->Style);
+      toporouter_netlist_t *nl = netlist_create(r, ((ConnectionType *)(((NetType *)(netlist->Net->data))->Connection->data))->menu->Name,
+                                                   ((ConnectionType *)(((NetType *)(netlist->Net->data))->Connection->data))->menu->Style);
       
       NET_LOOP(netlist);
       {
diff --git a/src/undo.c b/src/undo.c
index 21b86f4..843527c 100644
--- a/src/undo.c
+++ b/src/undo.c
@@ -164,7 +164,7 @@ UndoListType, *UndoListTypePtr;
  * some local variables
  */
 static DataTypePtr RemoveList = NULL;	/* list of removed objects */
-static UndoListTypePtr UndoList = NULL;	/* list of operations */
+static GList *UndoList = NULL;	/* list of operations */
 static int Serial = 1,		/* serial number */
   SavedSerial;
 static size_t UndoN, RedoN,	/* number of entries */
@@ -202,13 +202,15 @@ static int PerformUndo (UndoListTypePtr);
 /* ---------------------------------------------------------------------------
  * adds a command plus some data to the undo list
  */
-static UndoListTypePtr
+static UndoListType *
 GetUndoSlot (int CommandType, int ID, int Kind)
 {
-  UndoListTypePtr ptr;
+  UndoListType *ptr;
   void *ptr1, *ptr2, *ptr3;
   int type;
   static size_t limit = UNDO_WARNING_SIZE;
+  size_t size;
+  GList *iter;
 
 #ifdef DEBUG_ID
   if (SearchObjectByID (PCB->Data, &ptr1, &ptr2, &ptr3, ID, Kind) == NO_TYPE)
@@ -216,53 +218,59 @@ GetUndoSlot (int CommandType, int ID, int Kind)
 	     Kind);
 #endif
 
-  /* allocate memory */
-  if (UndoN >= UndoMax)
+  size = UndoN * sizeof (UndoListType);
+
+  /* ask user to flush the table because of it's size */
+  if (size > limit)
     {
-      size_t size;
+      limit = (size / UNDO_WARNING_SIZE + 1) * UNDO_WARNING_SIZE;
+      Message (_("Size of 'undo-list' exceeds %li kb\n"),
+               (long) (size >> 10));
+    }
 
-      UndoMax += STEP_UNDOLIST;
-      size = UndoMax * sizeof (UndoListType);
-      UndoList = (UndoListTypePtr) realloc (UndoList, size);
-      memset (&UndoList[UndoN], 0, STEP_REMOVELIST * sizeof (UndoListType));
+  /* free structures from the pruned redo list */
 
-      /* ask user to flush the table because of it's size */
-      if (size > limit)
-	{
-	  limit = (size / UNDO_WARNING_SIZE + 1) * UNDO_WARNING_SIZE;
-	  Message (_("Size of 'undo-list' exceeds %li kb\n"),
-		   (long) (size >> 10));
-	}
+  for (iter = g_list_nth (UndoList, UndoN);
+       iter != NULL; iter = g_list_next (iter), RedoN--)
+    {
+      ptr = iter->data;
+
+      switch (ptr->Type)
+        {
+        case UNDO_CHANGENAME:
+          free (ptr->Data.ChangeName.Name);
+          break;
+        case UNDO_REMOVE:
+          type =
+            SearchObjectByID (RemoveList, &ptr1, &ptr2, &ptr3, ptr->ID,
+                              ptr->Kind);
+          if (type != NO_TYPE)
+            {
+              DestroyObject (RemoveList, type, ptr1, ptr2, ptr3);
+            }
+          break;
+        default:
+          break;
+        }
+      g_slice_free (UndoListType, ptr);
     }
 
-  /* free structures from the pruned redo list */
+  /* Break off and free the pruned redo list */
+  iter = g_list_nth (UndoList, UndoN);
+  iter->prev->next = NULL;
+  iter->prev = NULL;
 
-  for (ptr = &UndoList[UndoN]; RedoN; ptr++, RedoN--)
-    switch (ptr->Type)
-      {
-      case UNDO_CHANGENAME:
-	free (ptr->Data.ChangeName.Name);
-	break;
-      case UNDO_REMOVE:
-	type =
-	  SearchObjectByID (RemoveList, &ptr1, &ptr2, &ptr3, ptr->ID,
-			    ptr->Kind);
-	if (type != NO_TYPE)
-	  {
-	    DestroyObject (RemoveList, type, ptr1, ptr2, ptr3);
-	  }
-	break;
-      default:
-	break;
-      }
+  g_list_free (iter);
 
   /* copy typefield and serial number to the list */
-  ptr = &UndoList[UndoN++];
+  ptr = g_slice_new0 (UndoListType);
+  UndoList = g_list_append (UndoList, ptr);
+
   ptr->Type = CommandType;
   ptr->Kind = Kind;
   ptr->ID = ID;
   ptr->Serial = Serial;
-  return (ptr);
+  return ptr;
 }
 
 /* ---------------------------------------------------------------------------
@@ -901,50 +909,35 @@ static bool
 UndoNetlistChange (UndoListTypePtr Entry)
 {
   NetlistChangeTypePtr l = & Entry->Data.NetlistChange;
-  unsigned int i, j;
+  GList *i, *j;
   LibraryTypePtr lib, saved;
 
   lib = l->lib;
   saved = l->old;
 
   /* iterate over each net */
-  for (i = 0 ; i < lib->MenuN; i++)
+  for (i = lib->Menu; i != NULL; i = g_list_next (i))
     {
-      if (lib->Menu[i].Name)
-	free (lib->Menu[i].Name);
+      LibraryMenuType *menu = i->data;
 
-      if (lib->Menu[i].directory)
-	free (lib->Menu[i].directory);
-
-      if (lib->Menu[i].Style)
-	free (lib->Menu[i].Style);
+      free (menu->Name);
+      free (menu->directory);
+      free (menu->Style);
 
       /* iterate over each pin on the net */
-      for (j = 0; j < lib->Menu[i].EntryN; j++) {
-	
-	if (lib->Menu[i].Entry[j].ListEntry)
-	  free (lib->Menu[i].Entry[j].ListEntry);
-	
-	if (lib->Menu[i].Entry[j].AllocatedMemory)
-	  free (lib->Menu[i].Entry[j].AllocatedMemory);
-	
-	if (lib->Menu[i].Entry[j].Template)
-	  free (lib->Menu[i].Entry[j].Template);
-	
-	if (lib->Menu[i].Entry[j].Package)
-	  free (lib->Menu[i].Entry[j].Package);
-	
-	if (lib->Menu[i].Entry[j].Value)
-	  free (lib->Menu[i].Entry[j].Value);
-	
-	if (lib->Menu[i].Entry[j].Description)
-	  free (lib->Menu[i].Entry[j].Description);
-	
+      for (j = menu->Entry; j != NULL; j = g_list_next (j)) {
+        LibraryEntryType *entry = j->data;
+
+        free (entry->ListEntry);
+        free (entry->AllocatedMemory);
+        free (entry->Template);
+        free (entry->Package);
+        free (entry->Value);
+        free (entry->Description);
       }
     }
 
-  if (lib->Menu)
-    free (lib->Menu);
+  free (lib->Menu);
 
   *lib = *saved;
 
@@ -1620,7 +1613,8 @@ void
 AddNetlistLibToUndoList (LibraryTypePtr lib)
 {
   UndoListTypePtr undo;
-  unsigned int i, j;
+  int i, j;
+  GList *ii, *jj;
   LibraryTypePtr old;
   
   if (!Locked)
@@ -1633,7 +1627,6 @@ AddNetlistLibToUndoList (LibraryTypePtr lib)
       undo->Data.NetlistChange.old = (LibraryTypePtr)malloc (sizeof (LibraryTypePtr));
       old = undo->Data.NetlistChange.old;
       old->MenuN = lib->MenuN;
-      old->MenuMax = lib->MenuMax;
       old->Menu = (LibraryMenuTypePtr)malloc (old->MenuMax * sizeof (LibraryMenuType));
       if (old->Menu == NULL)
 	{
@@ -1642,21 +1635,15 @@ AddNetlistLibToUndoList (LibraryTypePtr lib)
 	}
 
       /* iterate over each net */
-      for (i = 0 ; i < lib->MenuN; i++)
+      for (ii = liib->Menu, i = 0; ii != NULL; ii = g_liist_next (ii), i++)
 	{
-	  old->Menu[i].EntryN = lib->Menu[i].EntryN;
-	  old->Menu[i].EntryMax = lib->Menu[i].EntryMax;
+	  MenuType *menu = ii->data;
 
-	  old->Menu[i].Name = 
-	    lib->Menu[i].Name ? strdup (lib->Menu[i].Name) : NULL;
-	  
-	  old->Menu[i].directory = 
-	    lib->Menu[i].directory ? strdup (lib->Menu[i].directory) : NULL;
-	  
-	  old->Menu[i].Style = 
-	    lib->Menu[i].Style ? strdup (lib->Menu[i].Style) : NULL;
+	  old->Menu[i].EntryN = menu->EntryN;
+	  old->Menu[i].Name = menu->Name ? strdup (menu->Name) : NULL;
+	  old->Menu[i].directory = menu->directory ? strdup (menu->directory) : NULL;
+	  old->Menu[i].Style = menu->Style ? strdup (menu->Style) : NULL;
 
-      
 	  old->Menu[i].Entry = 
 	    (LibraryEntryTypePtr)malloc (old->Menu[i].EntryMax * sizeof (LibraryEntryType));
 	  if (old->Menu[i].Entry == NULL)
@@ -1666,39 +1653,19 @@ AddNetlistLibToUndoList (LibraryTypePtr lib)
 	    }
 	  
 	  /* iterate over each pin on the net */
-	  for (j = 0; j < lib->Menu[i].EntryN; j++) {
-
-	    old->Menu[i].Entry[j].ListEntry = 
-	      lib->Menu[i].Entry[j].ListEntry ? 
-	      strdup (lib->Menu[i].Entry[j].ListEntry) :
-	      NULL;
-
-	    old->Menu[i].Entry[j].AllocatedMemory = 
-	      lib->Menu[i].Entry[j].AllocatedMemory ? 
-	      strdup (lib->Menu[i].Entry[j].AllocatedMemory) :
-	      NULL;
-
-	    old->Menu[i].Entry[j].Template = 
-	      lib->Menu[i].Entry[j].Template ? 
-	      strdup (lib->Menu[i].Entry[j].Template) :
-	      NULL;
-
-	    old->Menu[i].Entry[j].Package = 
-	      lib->Menu[i].Entry[j].Package ? 
-	      strdup (lib->Menu[i].Entry[j].Package) :
-	      NULL;
-
-	    old->Menu[i].Entry[j].Value = 
-	      lib->Menu[i].Entry[j].Value ? 
-	      strdup (lib->Menu[i].Entry[j].Value) :
-	      NULL;
-
-	    old->Menu[i].Entry[j].Description = 
-	      lib->Menu[i].Entry[j].Description ? 
-	      strdup (lib->Menu[i].Entry[j].Description) :
-	      NULL;
-	    
-
+	  for (jj = menu->Entry, j = 0; jj != NULL; jj = g_list_next (jj), j++) {
+            LibraryEntryType *entry = jj->data;
+
+	    old->Menu[i].Entry[j].ListEntry =
+              entry->ListEntry ? strdup (entry->ListEntry) : NULL;
+	    old->Menu[i].Entry[j].AllocatedMemory =
+              entry->AllocatedMemory ? strdup (entry->AllocatedMemory) : NULL;
+	    old->Menu[i].Entry[j].Template =
+	      entry->Template ? strdup (entry->Template) : NULL;
+	    old->Menu[i].Entry[j].Package = entry->Package ? strdup (entry->Package) : NULL;
+	    old->Menu[i].Entry[j].Value = entry->Value ? strdup (entry->Value) : NULL;
+	    old->Menu[i].Entry[j].Description =
+              entry->Description ? strdup (entry->Description) : NULL;
 	  }
 	}
